File: lib/src/http/http_service_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import '../core/service_core.dart';
import 'auth_exceptions.dart';
import 'http_exceptions.dart';
import 'token.dart';
import 'token_vault.dart';

class HttpServiceClient implements ServiceClient {
  HttpServiceClient(this.config) : _client = http.Client();

  final ServiceClientConfig config;
  final http.Client _client;

  @override
  Future<ServiceResponse> send(ServiceRequest request) async {
    if (request.protocol != ServiceProtocol.http) {
      throw Exception('Unsupported protocol for HttpServiceClient');
    }

    final url = config.baseUrl.resolve(request.endpoint);
    final effectiveHeaders = <String, String>{
      'Content-Type': 'application/json',
      ...config.defaultHeaders,
      if (request.headers != null) ...request.headers!,
    };

    if (config.auth && Token.accessToken != null) {
      effectiveHeaders['Authorization'] = 'Bearer ${Token.accessToken}';
    }

    String? encodeBody(Object? body, {bool useEmptyObjectWhenNull = false}) {
      if (body == null) {
        return useEmptyObjectWhenNull ? jsonEncode(<String, dynamic>{}) : null;
      }
      if (body is String) return body;
      return jsonEncode(body);
    }

    Future<http.Response> doCall() {
      final method = request.method.toUpperCase();
      switch (method) {
        case 'GET':
          return _client
              .get(url, headers: effectiveHeaders)
              .timeout(config.timeout);
        case 'POST':
          return _client
              .post(
                url,
                headers: effectiveHeaders,
                body: encodeBody(request.body, useEmptyObjectWhenNull: true),
              )
              .timeout(config.timeout);
        case 'PATCH':
          return _client
              .patch(
                url,
                headers: effectiveHeaders,
                body: encodeBody(request.body),
              )
              .timeout(config.timeout);
        default:
          final req = http.Request(method, url)
            ..headers.addAll(effectiveHeaders);
          final encoded = encodeBody(request.body);
          if (encoded != null) {
            req.body = encoded;
          }
          return _client
              .send(req)
              .then(http.Response.fromStream)
              .timeout(config.timeout);
      }
    }

    try {
      var response = await doCall();

      final bool isLogin = request.endpoint.contains('/auth/login');
      if (isLogin && config.auth && response.statusCode == 200) {
        final m = jsonDecode(response.body) as Map<String, dynamic>;
        final access = m['access_token'] as String?;
        final expiresIn = (m['expires_in'] as num?)?.toInt();
        final rt = m['refresh_token'] as String?;

        if (access != null) {
          Token.accessToken = access;
        }
        if (expiresIn != null) {
          Token.accessExp = DateTime.now().add(Duration(seconds: expiresIn));
        }
        if (rt != null) {
          await TokenVault.saveRefresh(_kCurrentUser, rt);
        }

        return _toServiceResponse(response, decodedBody: m);
      }

      if (response.statusCode == 401 && config.auth) {
        final refreshEndpoint = _inferRefreshEndpoint(request.endpoint);
        final refreshed = await _tryRefresh(
          baseUrl: config.baseUrl,
          refreshEndpoint: refreshEndpoint,
          client: _client,
          timeout: config.timeout,
        );
        if (refreshed) {
          if (Token.accessToken != null) {
            effectiveHeaders['Authorization'] = 'Bearer ${Token.accessToken}';
          }
          response = await doCall();
        }
      }

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return _toServiceResponse(response);
      }

      if (response.statusCode == 401 && config.auth) {
        throw AuthReLoginException();
      }

      // Log error response body for debugging
      stderr.writeln('HTTP Error Response:');
      stderr.writeln('  Status: ${response.statusCode}');
      stderr.writeln('  Body: ${response.body}');
      
      // Parsear response body si es JSON
      Map<String, dynamic>? responseData;
      try {
        responseData = jsonDecode(response.body) as Map<String, dynamic>?;
      } catch (_) {
        // Si no se puede parsear, dejar como null
      }
      
      throw HttpClientException(
        statusCode: response.statusCode,
        message: request.errorMessage,
        response: responseData,
      );
    } catch (e) {
      if (e is AuthReLoginException) rethrow;
      if (e is HttpClientException) rethrow;

      stderr.writeln('HTTP Client Error: $e');
      throw Exception('${request.errorMessage}: [Connection error] - $e');
    }
  }

  @override
  Future<void> close() async {
    _client.close();
  }

  ServiceResponse _toServiceResponse(
    http.Response response, {
    dynamic decodedBody,
  }) {
    final rawBody = response.body;
    final data =
        decodedBody ?? (rawBody.isNotEmpty ? jsonDecode(rawBody) : null);
    return ServiceResponse(
      statusCode: response.statusCode,
      headers: response.headers,
      rawBody: rawBody,
      data: data,
    );
  }

  String _inferRefreshEndpoint(String endpoint) {
    var refreshEndpoint = 'auth/refresh';
    if (endpoint.contains('/')) {
      final parts = endpoint.split('/');
      if (parts.isNotEmpty && parts[0].isNotEmpty) {
        refreshEndpoint = '${parts[0]}/auth/refresh';
      }
    }
    return refreshEndpoint;
  }
}

const String _kCurrentUser = 'current_user';

Future<bool> _tryRefresh({
  required Uri baseUrl,
  required String refreshEndpoint,
  required http.Client client,
  required Duration timeout,
}) async {
  final rt = await TokenVault.readRefresh(_kCurrentUser);
  if (rt == null) return false;

  final url = baseUrl.resolve(refreshEndpoint);
  try {
    final r = await client
        .post(
          url,
          headers: const {'Content-Type': 'application/json'},
          body: jsonEncode({'refresh_token': rt}),
        )
        .timeout(timeout);

    if (r.statusCode == 200) {
      final m = jsonDecode(r.body) as Map<String, dynamic>;
      final access = m['access_token'] as String?;
      final expiresIn = (m['expires_in'] as num?)?.toInt();
      final newRt = m['refresh_token'] as String?;

      if (access == null) return false;

      Token.accessToken = access;
      if (expiresIn != null) {
        Token.accessExp = DateTime.now().add(Duration(seconds: expiresIn));
      }
      if (newRt != null) {
        await TokenVault.saveRefresh(_kCurrentUser, newRt);
      }

      return true;
    }

    return false;
  } catch (e) {
    stderr.writeln('Token refresh error: $e');
    return false;
  }
}

File: code/py/src/service_client/http/http_service_client.py
"""HttpServiceClient — implementación de ServiceClient usando httpx.

Usa httpx.AsyncClient para peticiones HTTP con timeout configurable.
Soporta GET, POST, PATCH, PUT, DELETE y cualquier otro método HTTP.
"""

from __future__ import annotations

import json
import logging
from urllib.parse import urljoin

import httpx

from ..core.service_client import (
    ServiceClientConfig,
    ServiceProtocol,
    ServiceRequest,
    ServiceResponse,
)
from .http_client_exception import HttpClientException

logger = logging.getLogger(__name__)


class HttpServiceClient:
    """ServiceClient implementation backed by httpx.AsyncClient.

    El parámetro `client` es opcional y existe exclusivamente para inyección en tests.
    """

    def __init__(
        self,
        config: ServiceClientConfig,
        *,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self._config = config
        self._client = client or httpx.AsyncClient(
            timeout=httpx.Timeout(config.timeout),
        )

    async def send(self, request: ServiceRequest) -> ServiceResponse:
        """Envía un request HTTP y retorna un ServiceResponse en caso de 2xx.

        Lanza HttpClientException para respuestas 4xx/5xx.
        Lanza RuntimeError para errores de conexión.
        """
        if request.protocol != ServiceProtocol.HTTP:
            raise RuntimeError("Unsupported protocol for HttpServiceClient")

        url = urljoin(self._config.base_url, request.endpoint)
        effective_headers = {
            "Content-Type": "application/json",
            **self._config.default_headers,
            **(request.headers or {}),
        }

        encoded_body = self._encode_body(
            request.body,
            is_post=request.method.upper() == "POST",
        )

        try:
            response = await self._client.request(
                method=request.method.upper(),
                url=url,
                headers=effective_headers,
                content=encoded_body,
            )

            raw_body = response.text

            if 200 <= response.status_code < 300:
                data = json.loads(raw_body) if raw_body else None
                return ServiceResponse(
                    status_code=response.status_code,
                    headers=dict(response.headers),
                    raw_body=raw_body,
                    data=data,
                )

            logger.debug("HTTP Error Response:")
            logger.debug("  Status: %d", response.status_code)
            logger.debug("  Body: %s", raw_body)

            response_data: dict[str, object] | None = None
            try:
                response_data = json.loads(raw_body)
            except (json.JSONDecodeError, ValueError):
                pass  # Response body is not valid JSON — leave as None

            raise HttpClientException(
                status_code=response.status_code,
                message=request.error_message,
                response=response_data,
            )

        except HttpClientException:
            raise

        except Exception as exc:
            logger.debug("HTTP Client Error: %s", exc)
            raise RuntimeError(
                f"{request.error_message}: [Connection error] - {exc}"
            ) from exc

    async def close(self) -> None:
        """Cierra el cliente HTTP subyacente."""
        await self._client.aclose()

    @staticmethod
    def _encode_body(body: object | None, *, is_post: bool) -> str | None:
        """Codifica el body como JSON string.

        POST envía '{}' cuando body es None (B-04).
        Si body ya es string, se pasa tal cual sin doble codificación (B-05).
        """
        if body is None:
            return "{}" if is_post else None
        if isinstance(body, str):
            return body
        return json.dumps(body)
